xarray/core/combine.py

```diff
@@ -456,107 +146,372 @@ def _combine_nd(combined_ids, concat_dims, data_vars='all',
+        combined_ids = _combine_all_along_first_dim(combined_ids,
+                                                    dim=concat_dim,
+                                                    data_vars=data_vars,
+                                                    coords=coords,
+                                                    compat=compat)
     combined_ds = list(combined_ids.values())[0]
     return combined_ds
 
 
-def _auto_combine_all_along_first_dim(combined_ids, dim, data_vars,
-                                      coords, compat):
+def _combine_all_along_first_dim(combined_ids, dim, data_vars, coords, compat):
+
     # Group into lines of datasets which must be combined along dim
     # need to sort by _new_tile_id first for groupby to work
     # TODO remove all these sorted OrderedDicts once python >= 3.6 only
     combined_ids = OrderedDict(sorted(combined_ids.items(), key=_new_tile_id))
     grouped = itertools.groupby(combined_ids.items(), key=_new_tile_id)
 
+    # Combine all of these datasets along dim
     new_combined_ids = {}
     for new_id, group in grouped:
         combined_ids = OrderedDict(sorted(group))
         datasets = combined_ids.values()
-        new_combined_ids[new_id] = _auto_combine_1d(datasets, dim, compat,
-                                                    data_vars, coords)
+        new_combined_ids[new_id] = _combine_1d(datasets, dim, compat,
+                                               data_vars, coords)
     return new_combined_ids
 
 
-def vars_as_keys(ds):
-    return tuple(sorted(ds))
-
+def _combine_1d(datasets, concat_dim, compat='no_conflicts', data_vars='all',
+                coords='different'):
+    """
+    Applies either concat or merge to 1D list of datasets depending on value
+    of concat_dim
+    """
 
-def _auto_combine_1d(datasets, concat_dim=_CONCAT_DIM_DEFAULT,
-                     compat='no_conflicts',
-                     data_vars='all', coords='different'):
-    # This is just the old auto_combine function (which only worked along 1D)
     if concat_dim is not None:
-        dim = None if concat_dim is _CONCAT_DIM_DEFAULT else concat_dim
-        sorted_datasets = sorted(datasets, key=vars_as_keys)
-        grouped_by_vars = itertools.groupby(sorted_datasets, key=vars_as_keys)
-        concatenated = [_auto_concat(list(ds_group), dim=dim,
-                                     data_vars=data_vars, coords=coords)
-                        for id, ds_group in grouped_by_vars]
+        try:
+            combined = concat(datasets, dim=concat_dim, data_vars=data_vars,
+                              coords=coords)
+        except ValueError as err:
+            if "encountered unexpected variable" in str(err):
+                raise ValueError("These objects cannot be combined along the "
+                                 "dimension {concat_dim} using only "
+                                 "xarray.concat, you must use "
```

Comment:
I'm not sure I understand this error message. This is used inside `combine_auto` or `auto_combine` or `combine_manual`, right? It seems a little confusing to mention `xarray.concat` when users may not have mentioned it explicitly.

Reply:
This is the most subtle part of the whole PR. It's covering the case where you want to combine datasets which require both concatenation and merging. For example (from `test_manual_combine_but_need_auto_combine`):

```python
objs = [Dataset({'x': [0, 1]}), Dataset({'x': [2], 'wall': [0]})]
with raises_regex(ValueError, 'cannot be combined'):
    combine_manual(objs, concat_dim='x')
```

You can't use `combine_manual` for this case because you can't use either `concat` or `merge` along the first (and only) axis of the `objs` array, you have to use a combination of both. However `combine_manual` works by applying just one of those operations to each axis of the input object array. The error message is telling the user to use `combine_auto` instead, because that function first groups all the variables according to their dims before concatenation, then merges them back together, so could handle a case like this.

This message is covering the edge case we discussed [here](https://github.com/pydata/xarray/pull/2553#issuecomment-445419876).

> This is used inside combine_auto or auto_combine or combine_manual, right?

This internal function (`_combine_1d`) is used inside all of those, but this error should only ever be raised by `combine_manual`, because only `combine_manual` could end up passing a combination of datasets like this to `_combine_1d`.

> It seems a little confusing to mention xarray.concat when users may not have mentioned it explicitly.

The fact we're using `concat` is supposed to be understood from the description of `combine_manual`. I agree that this is confusing - perhaps the message should just say something like "can't handle this combination of variables, instead use either `combine_auto` or just do it manually with `concat` and `merge`"?

Reply:
> You can't use `combine_manual` for this case because you can't use either `concat` or `merge` along the first (and only) axis of the `objs` array, you have to use a combination of both. However `combine_manual` works by applying just one of those operations to each axis of the input object array. The error message is telling the user to use `combine_auto` instead, because that function first groups all the variables according to their dims before concatenation, then merges them back together, so could handle a case like this.

In principle, I think you could use `align()` in this case (possibly with `exclude` to skip some dimensions), and then use `combine_manual` afterwards. 

> The fact we're using `concat` is supposed to be understood from the description of `combine_manual`.

If possible, I think it would be best to treat the usage of `concat` and `merge` inside `combine` as an implementation detail.

> I agree that this is confusing - perhaps the message should just say something like "can't handle this combination of variables, instead use either `combine_auto` or just do it manually with `concat` and `merge`"?

Yes, that would be better -- but also add `align` to that list of functions to use :)

Reply:
> In principle, I think you could use align() in this case (possibly with exclude to skip some dimensions), and then use combine_manual afterwards.

If this comes up in future then I guess we could add it as an example to the documentation.

> If possible, I think it would be best to treat the usage of concat and merge inside combine as an implementation detail.

That works for `combine_auto`, but `combine_manual` was always supposed to be the more explicit version, and the docs for that function do explain how it uses `concat` and `merge`.

> Yes, that would be better -- but also add align to that list of functions to use :)

Okay cool, I will do that now